Differentiate materials in DAGMC universes

bam241 opened the pull request:
<!--
If you are a first-time contributor to OpenMC, please have a look at our
contributing guidelines:
https://github.com/openmc-dev/openmc/blob/develop/CONTRIBUTING.md
-->

# Description

This allows to differentiate materials in DAGMC universe when using the same DAGMC geometry to fill different CG cells

This provides a mechanism to overload material assignment embedded in the DAGMC geometry by providing a dict `{ref_mat_name:[mat_instances]}`
This allows enrich the `Model.differentiate_depletable_material()` method from the Python API, to allows automatic differentiation of the depletable materials of a DAGMC geometry inserted multiple times.

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)


this should address #2923 

_This work has been sponsored by [NAAREA](https://www.naarea.fr/en)._

pshriwise commented on openmc/cell.py:
```diff
@@ -780,3 +782,83 @@ def from_xml_element(cls, elem, surfaces, materials, get_universe):
         return c
+
+
+class DAGSpeudoCell(Cell):
```
Let's call this `DAGMCCell` instead.

pshriwise commented on src/dagmc.cpp:
```diff
@@ -72,6 +72,28 @@ DAGUniverse::DAGUniverse(pugi::xml_node node)
+      std::string mat_ref_assignment = attr.name();
+
+      // Get mat name for each assignement instances
+      std::stringstream iss {attr.value()};
```
Let's use the `split` function from `string_utils.h` here.

pshriwise replied:
A reorganization of the XML node might make this easier also. We can revisit that later.

pshriwise commented on openmc/universe.py:
```diff
@@ -823,12 +825,16 @@ def __init__(self,
         self.filename = filename
         self.auto_geom_ids = auto_geom_ids
         self.auto_mat_ids = auto_mat_ids
+        self.mat_assignment = mat_assignment
```
Can we call this `material_overrides`? 

Also, since we'll need to add a doc string for this eventually, can you add that to `DAGMCUniverse` with a description of the data structure keys/values and it's purpose?

pshriwise commented on openmc/universe.py:
```diff
@@ -871,6 +877,12 @@ def auto_mat_ids(self, val):
         self._auto_mat_ids = val
 
+    @property
+    def material_assignment(self):
```
These values aren't going to exclusively be material assignments. The `NAME` tag in the `.h5m` file also contains boundary conditions, temperatures, etc.

https://pshriwise.github.io/dagmc-docs/filespec/index.html

pshriwise replied:
Though it looks like this could be removed in favor of the `material_names` property below?

pshriwise commented on openmc/universe.py:
```diff
@@ -1117,3 +1178,64 @@ def _partial_deepcopy(self):
+        # If the Cell is in the Universe's list of Cells, delete it
+        self._cells.pop(cell.id, None)
+
+    def clear_cells(self):
```
This is alread provided on `UniverseBase`. Is there a reason it is duplicated here?

bam241 replied:
This is part of `Universe`, not `UniverseBase` directly.
As `DAGMCUniverse` inherit from `UniverseBase` I as unsure of the best implementation.

I can move all of them in `UniverseBase` so they both inherit from the methods.

pshriwise replied:
You're right, they aren't inherited methods on the `Universe` class, but if we're implementing the same method in two places it would be good to consider whether or not they could be placed on the `UniverseBase` class, even as abstract methods so we have a consistent interface.

pshriwise commented on include/openmc/cell.h:
```diff
@@ -349,6 +348,12 @@ class Cell {
 
   vector<int32_t> offset_; //!< Distribcell offset table
+
+  const GeometryType& geom_type() const { return geom_type_; }
```
```suggestion
  // Accessors
  const GeometryType& geom_type() const { return geom_type_; }
```

pshriwise commented on openmc/geometry.py:
```diff
@@ -404,6 +404,23 @@ def get_all_nuclides(self) -> list[str]:
             all_nuclides |= set(material.get_nuclides())
         return sorted(all_nuclides)
+    
+    def get_all_dag_universes(self) -> typing.Dict[int, openmc.DAGMCUniverse]:
```
```suggestion
    def get_all_dagmc_universes(self) -> typing.Dict[int, openmc.DAGMCUniverse]:
```

pshriwise commented on openmc/universe.py:
```diff
@@ -823,12 +825,16 @@ def __init__(self,
         self.auto_geom_ids = auto_geom_ids
         self.auto_mat_ids = auto_mat_ids
+        self.mat_assignment = mat_assignment
+        self._num_instances = 0
```
Thinking about this more I think we should place this attribute on the `DAGMCPseudoCell`/`DAGMCCell` class. It may be a little redundant, but it will reduce the amount of DAGMC-specific code at higher levels of the `Model` and `Geometry` classes.

pshriwise commented on openmc/model/model.py:
```diff
@@ -1058,8 +1130,59 @@ def differentiate_depletable_mats(self, diff_volume_method: str):
+                            cell.fill = mat.clone()
+                        
+        dag_verse_mats = []
+        for dag_verse in self.geometry.get_all_dag_universes().values():
```
I wonder if some adjustments to the `DAGMC*` geometry classes in the Python API would let us use the same logic as CSG universes above.

pshriwise replied:
Perhaps the resulting `deplete_mat_dict` which is built up here would be a barrier to that, but I wonder if we couldn't built that on-the-fly when we're exporting the DAGMCUniverse XML based on material assignments present on the `DAGMCCell` objects stored on the `DAGMCUniverse`?

bam241 commented:
@pshriwise I think I addressed most of you comments (on the python side, didn't find the time to touch the 1 or 2 comments about the cpp..).

I didn't revolve the conversation so you can keep track of them.

One remaining question is: shall I "refactor" `BaseUniverse`/`Universe`/`DAGMCUniverse` ?

I am also happy to implement a DAGMC.py if you fell it is needed, while I am at it, it is probably less work than to go back at it later :)

let me know

pshriwise commented:
> One remaining question is: shall I "refactor" `BaseUniverse`/`Universe`/`DAGMCUniverse` to avoid redundancy as much as possible?

I think adding abstract methods for the cell modification methods would be good as a start. I do wonder if we might be able to rely on the same implementation in `Universe` and `DAGMCUniverse` for those methods. It might be worth a go to see if the implementation gets messy handling both the `Cell` and `DAGMCCell` types. We could make a base class, `CellBase` but I have a feeling that would add a level of inheritance that we don't really need b/c, as the current class name suggests, the DAGMC cells are pseudo cells.
 
> I am also happy to implement a DAGMC.py if you fell it is needed, while I am at it, it is probably less work than to go back at it later :)

If you're up for it I think that would be a good improvement from a codebase organization standpoint! Thanks for being willing to do so ❤️

bam241 reviewed (commented):
just posting some comments that didn't go through....

bam241 commented:
@pshriwise I think I have addressed all your comments.
(factorisation of the different Universe methods between `UniverseBase`, `Universe` and `DAGMCUniverse`)
and created a specific `dagmc.py` that contains `DAGMCUniverse` and `DAGMCCell`

You mentioned something about improving the xml nodes, shall we go back to it ?


There is also an problem I spotted when doing my tests.
When using `model.differentiate_depletable_mat("match cell")`, the volume of the cell is applied to each materials,
When using the DAGMCUniverse to fill multiple `Cells` the `DAGMCCell`s have multiple instances, and the volume of them is:
`DAGMCCell.Volume = Real_Cell_volume * DAGMCCell._n_instance`
(where Real_Cell_volume is the actual volume of the DAGMCCell)


I am guessing this is not a specific DAGMC problem:
if we use any Universe (DAGMC or not) to fill multiple `Cells` of the Geometry, one will end up with the same issue.
Shall we do something about this ? Add a warning ?


Also, when using `model.differentiate_mats` and running new volume calculation I need to run `model.init_lib()` multiple time, to resync the cpp.
Shall I add a method to sync the cpp with the geometry/model change in python ?
and call it if the cpp_lib is initialised in the method that would change the model?

i.e.:
to run a volume calculation on the DAGMCCells I need:
1. definition of the geometry in Python
2. model.init_lib() (to recover the information of the DAGMCCells)
3. definition of the Volume calculation on the DAGMCCell
4. model.init_lib() (to update the xml with the new volume calculation information)
5. model.calculate_volumes()
...

bam241 commented:
@pshriwise following up on the CI issu I mentioned Friday:
as you can see in the commit history:
[3197ee4](https://github.com/openmc-dev/openmc/pull/3056/commits/3197ee4d3616d3714018aadebece9e6b753d79a9) passed, 
but [3a7c19c](https://github.com/openmc-dev/openmc/pull/3056/commits/3a7c19c23846d50beccfcbebf2f284efceb8e05e) didn't.

the only difference is:
openmc/model/model.py
```diff
@@ -1050,10 +1050,6 
            volume equally between the new materials, 'match cell' sets the
            volume of the material to volume of the cell they fill.
        """
-       # if diff_volume_method not in ['divide equally', 'match cell']:
-       #     raise ValueError(
-       #         f"diff_volume_method must be 'divide equally' or 'match cell', "
-       #         f"not '{diff_volume_method}'")
        # Count the number of instances for each cell and material
        self.geometry.determine_paths(instances_only=True)
```

azimgivron commented on include/openmc/cell.h:
```diff
@@ -349,6 +348,13 @@ class Cell {
+
+  // Accessors
+  const GeometryType& geom_type() const { return geom_type_; }
+  GeometryType& geom_type() { return geom_type_; }
```
```suggestion
```

azimgivron commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,498 @@
+        self._parent_universe = universe.id
+
+    def boundingbox(self):
+        print("Warning: Bounding box is not available for cells in a DAGMC universe.")
```
```suggestion
        raise NotImplementedError("Bounding box is not available for cells in a DAGMC universe")
```

pshriwise reviewed (changes requested):
Some line comments and a couple of design questions. The main one being how material overrides are handled if applied to multiple cells.

pshriwise commented:
I wanted to note that the ability to differentiate materials in the `Model.differentiate_mats`  without applying material volumes added in this PR might help with those looking to determine the volume of cell instances (albeit in a roundabout way by performing a material volume calculation afterward)

https://github.com/openmc-dev/openmc/issues/2367

bam241 commented on openmc/lib/dagmc.py:
```diff
@@ -0,0 +1,42 @@
+_dll.openmc_get_dagmc_cell_ids.errcheck = _error_handler
+
+
+def get_dagmc_cell_ids(volume_id, n_cells):
```
it is possible, but we need the number of cells to instantiate the array to collect those ids to the right size, that means we need 2 calls to `_dll.openmc_get_dagmc_cell_ids(`
one to get the size, a second to fill the cell_ids array.

Passing a dynamic array did't work, so I had to use a fixed size array.


So what I can do:
```python
openmc.dagmc.sync_dagmc_cells() {
[...]
  cell_ids, found_cells = openmc.lib.dagmc.get_dagmc_cell_ids()
  assert found_cells == self.n_cells
[...]
}
```
with :
```python
openmc.lib.dagmc.get_dagmc_cell_ids(){
[...]
  # Get number of cells
  n = c_size_t()
  _dll.openmc_get_dagmc_cell_ids(volume_id, None, n)
  
  # Allocate array of cell ids 
  cell_ids = np.empty(n.value, dtype=np.int32)
  _dll.openmc_get_dagmc_cell_ids(volume_id, cell_ids.ctypes.data_as(POINTER(c_int32)), n )
[...]  
}

pshriwise replied:
Fair points, I'm thinking that it would probably make sense to have a separate CAPI call (something like `openmc_dagmc_universe_get_num_cells`) to get the number of cells from a DAGMC universe. The cross-over of information from the pure Python object to the Python object linked to a C++ object via `openmc.lib` is what feels off to me here.

bam241 replied:
good idea !
I'll do that !

bam241 commented on openmc/model/model.py:
```diff
@@ -1026,38 +1049,51 @@ def differentiate_depletable_mats(self, diff_volume_method: str):
-                if mat.depletable and mat.num_instances > 1])
-
-        if diff_volume_method == 'divide equally':
+            [
```
auto format, reverting this.

bam241 commented:
@pshriwise, 
I think I have addressed all your comments.

the only remaining is the reference the materials per id in the override_material dict

pshriwise commented on openmc/universe.py:
```diff
@@ -53,6 +53,10 @@ def __repr__(self):
     def name(self):
         return self._name
 
+    @property
```
True, there is no setter, but this property helps to ensure that the `Universe`'s contained cells are modified with the `add_cell`/`remove_cell` methods rather than being modified directly.

pshriwise replied:
Actually, I suppose those methods can be worked around by modifying the object returned by the property, but it does avoid setting it to something like

```python
universe.cells = [openmc.Cell(), openmc.Cell()]
```

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,576 @@
+                      Warning)
+        return None
+
+    def plot(self, *args, **kwargs):
```
I think this method is possible to use with DAGMC universes as they are?

pshriwise replied:
Noting that we should probably add a test (outside this PR that exercises this functionality if true)

bam241 replied:
`plot()` is part of `Universe`, Not `UniverseBase`
Shall I move the `Plot()` implementation in `UniverseBase` and the warning in `DAGMCUniverse()` ?

bam241 replied:
The `Plot` method your are pointing to is the one from `DAGMCCell`

The connection I implement, only recovers ID and fill, this is why a lot of methods are overridden from openmc.Cell inheritance.

pshriwise commented on openmc/lib/dagmc.py:
```diff
@@ -0,0 +1,62 @@
+_dll.openmc_dagmc_universe_get_num_cells.errcheck = _error_handler
+
+
+def get_dagmc_cell_ids(dagmc_id):
```
I know this would be a significant change, and one that I'm open to leaving to a subsequent PR, but I wonder if a `DAGMCUniverse` class implemented here could benefit from a lot of existing functionality of the `Universe` class.

pshriwise replied:
Fantastic! That will be rather nifty I think.

pshriwise commented on src/dagmc.cpp:
```diff
@@ -206,15 +225,24 @@ void DAGUniverse::init_geometry()
+            instance_material_overrides.find(mat_str) !=
+              instance_material_overrides.end()) {
+
+          for (auto mat_str_instance :
```
I think this question still stands? If I'm missing something can you point to the response or code change that resolves it?

bam241 replied:
I might have miss understood you, if so sorry for marking this as resolved.

I added the test before the for loop.
L236-242

If the material overrides apply to multiple cells, that means that those cells shared the same initial material assignement in the same DAGMCgeometry.



I think I see where the problem is, this works to differenciate materials in case of multiple insertion of the same DAGMC Geometry, this does not aim to differenciate assignement of a single material to multiple cells of the same DAGMC Universe.

pshriwise replied:
Yeah exactly. If the multiple DAGMC cells uses the same material and those cells have multiple instances, it isn't clear to which cell the material overrides will apply. It seems as though it would be the same for both cells, resulting in the same material usage for those cells in each use of the DAGMC universe. In the end I don't think this will accomplish material differentiation as desired.

bam241 replied:
That was the intended behaviour, 
Not to change the number of materials used in a DAGMC geometry but to replace the assigned materials each time the DAGMC is inserted again.

So if two cells share the same material assignment in the h5 file, the overriding material will be apply to both cells.

To go back to the previous loop, 
mat_A is assigned to Cell1 and Cell2 that are part of a DAGMC Geometry 
so we have:
`Cell1.material_[0] = mat_A`
`Cell2.material_[0] = mat_A`

If we insert the DAGMC geometry 3 times and we have a override element such as: `mat_A:[mat_A_1, mat_A_2, mat_A_3]`

So
`Cell1.material_[0] = mat_A_1`
`Cell1.material_[1] = mat_A_2`
`Cell1.material_[2] = mat_A_3`
`Cell2.material_[0] = mat_A_1`
`Cell2.material_[1] = mat_A_2`
`Cell2.material_[2] = mat_A_3`

I am happy to discuss if we want to have a different behaviour

pshriwise replied:
I was just looking at the `differentiate_depletable_mats` function again and it appears to me that we'd expect something more like.

```
Cell1.material_[0] = mat_A_0
Cell1.material_[1] = mat_A_1
Cell1.material_[2] = mat_A_2

Cell2.material_[0] = mat_A_3
Cell2.material_[1] = mat_A_4
Cell2.material_[2] = mat_A_5
```

Even if those two cells are in the same universe (a DAGMC universe in this case). This is because a clone of the original material is created each application of the material in a cell (including for each cell instance). 

As I see it, the problem with what is currently implemented it is that we'll get what you've described above, which is inconsistent with what would happen for CSG geometry. The question to me is, how does Cell2 know where to start reading the materials in the list provided by `material_overrides`? This is one reason I think it might make more sense to use cells/cell IDs as the keys of the `material_overrides` dictionary.

Separately, I do think that it would be more flexible and broadly applicable to use DAGMC cell IDs as the keys of the `material_override` object on `DAGMCUniverses`. This would allow someone to easily override a DAGMC material assignment outside of a depletion use case.

pshriwise replied:
> I am happy to discuss if we want to have a different behaviour, but I can see why differentiate material in a way that differentiation of a material assigned to multiple volumes of the DAGMC geometry, could be useful, but it could also be dangerous.

Can you elaborate on this a bit?

bam241 replied:
Discussing with @nplinden we were thinking that allowing to "change" the material assignment done in the h5m file could lead to problem (i.e. differentiating material within the same geometry) while we only aimed to differentiate materials when importing multiple time the same DAGMC geomretry.


It took me a while to understand what you were saying (sorry), that as we follow the existing behaviour of `differentiate_depletable_mat`
You are absolutely right!

I see 2 possible implementation there:
- as you mention keep the override reference by name, and use the cell_Id+instance in the cloned mat and use this as key for the new assignment, 

- if it is preferable to reference the overrides per id, when assigning the materials, one could make a local copy of the overrides map and pop the elements from the override list while the are reassigned.
re using the previous example:
make a local copy of:

`override = mat_A:[mat_A_0, mat_A_1, mat_A_2, mat_A_3, mat_A_4, mat_A_5]`

```
override_copy = mat_A:[mat_A_0, mat_A_1, mat_A_2, mat_A_3, mat_A_4, mat_A_5]
[...]
Cell1.material_[0] = mat_A[0]
override_copy = mat_A:[mat_A_1, mat_A_2, mat_A_3, mat_A_4, mat_A_5]
Cell1.material_[1] = mat_A[0]
override_copy = mat_A:[mat_A_2, mat_A_3, mat_A_4, mat_A_5]
Cell1.material_[2] = mat_A[0]
override_copy = mat_A:[mat_A_3, mat_A_4, mat_A_5]
[...]
Cell1.material_[3] = mat_A[0]
override_copy = mat_A:[mat_A_4, mat_A_5]
Cell1.material_[4] = mat_A[0]
override_copy = mat_A:[mat_A_5]
Cell1.material_[5] = mat_A[0]
override_copy = mat_A:[]
[...]

```

pshriwise replied:
> Discussing with @nplinden we were thinking that allowing to "change" the material assignment done in the h5m file could lead to problem (i.e. differentiating material within the same geometry) while we only aimed to differentiate materials when importing multiple time the same DAGMC geomretry.

Ah I see. Yes, for this particular use case I can see how that might be confusing.


>  - as you mention keep the override reference by name, and use the cell_Id+instance in the cloned mat and use this as key for the new assignment,

This is close to what I had in mind. I think in the `material_overrides` dictionary each entry can use a cell ID as a key. The value of this entry can be either a list of length one (the material override applies to all cell instances) or an `Iterable` with a size equal to the number of cell instances -- this can be verified on the Python or C++ side (or both). This is how the `openmc.Cell.fill` attribute is handled and checked at runtime on the C++ side. That thought leads me back to wondering if it might be viable to add a `fill` attribute to the `DAGMCCell` object. Those 'fill' attributes can be written with the DAGMCUniverse XML node if present.

bam241 replied:
ok so we need to enable 2 override modes :
- 1 done automatically with `differentiate_depletable_mat`, referencing `DAGMCCell.id` to a list of mats (id?) corresponding to each instance of the Cell
- 1 when user manually override materials assignment of the DAGMC h5 file, referencing  the original material_name to a list of mats (id?) corresponding to each instance of the DAGMC universe

bam241 replied:
what we can do, if check if the `key` is in the original mat assignment and switch it to `id` search if not.

I understood you'd prefer if we reference the materials used for the override by id not by mat ?
so having something like:
`{key:[mat_id_1,...,mat_id_n}`

correct ?

pshriwise replied:
> ok so we need to enable 2 override modes :

Something like that, yeah. I think of this as two different use cases that we're trying to support with the same data structure.

I think I understand better now what your concern is regarding "dangerous" behavior. In theory it would be possible for a user to call `diff_burnable_mats` and then change the overrides dictionary entry. I'd propose two solutions for this (both involve checking the new assignments against the original assignment in the DAGMC file). 

  a. Perform the check during the `diff_burnable_mats` method as a way of verifying the updated material assignments
  b. Perform the check upon export of the DAGMC universe to XML and display a warning or raise an exception 

Personally, I like option a. I know it leaves more room for error if the user makes manual modifications after the call to `diff_burnable_mats`, but historically the design of OpenMC has favored flexibility in these cases.

Not trying to be pedantic or nit-picky. Mainly I want to make sure we're on the same page so your time isn't wasted if we aren't in agreement.

pshriwise replied:
Going back to data structures, what do you think of representing the overrides in a `DAGMCCell.fill` attribute?

bam241 replied:
> Personally, I like option a. I know it leaves more room for error if the user makes manual modifications after the call to `diff_burnable_mats`, but historically the design of OpenMC has favored flexibility in these cases.

When talking about user manual override of the material assignment I was not referring to user manual modification after the call to `diff_burnable_mats`
but simply a user wanting to override some materials of the DAGMC geometry.
writing manually the `DAGMCUniverse.material_overrides` or using `DAGMCUniverse.add_material_override`

I think it is a nice feature to be able to "replace" material from the original DAGMC geometry by mapping the original name with the news materials (for either unique or multiple DAGMC occurrence). It is also nice that it is 100% Python so it does not require to have the `openmc.lib` enabled.

I also like the "vision" you have, when using `diff_burnable_mats` to create a mapping  between DAGMCCell.id and the materials for each instances of this cell.

IMO, we need to enable both, and find a way to pass that information correctly to the cpp.

bam241 replied:
> Going back to data structures, what do you think of representing the overrides in a `DAGMCCell.fill` attribute?

Do you mean when using the `diff_burnable_mats` ?
because we already write the override in the `cell.fill`.
it is then used to generate the `material_override` node in the xml.

As to have `find_buranle_mats` only works with `DAGMCCell` when `openmc.lib` is enable, we could pass back the information directly to the cpp DAGMCCell. 
Or add a method to pass it.

Is that what you meant ?

pshriwise replied:
> When talking about user manual override of the material assignment I was not referring to user manual modification after the call to `diff_burnable_mats` but simply a user wanting to override some materials of the DAGMC geometry writing manually the `DAGMCUniverse.material_overrides` or using `DAGMCUniverse.add_material_override`

 > I think it is a nice feature to be able to "replace" material from the original DAGMC geometry by mapping the original name with the news materials (for either unique or multiple DAGMC occurrence). 

> IMO, we need to enable both, and find a way to pass that information correctly to the cpp.

I think we're in agreement that we can and should support both use cases of material overrides we've outlined above. My main point was that I think we can use the same override data structure (be it with a `dict` on the `DAGMCUniverse` or using the `DAGMCCell.fill` attribute).

> It is also nice that it is 100% Python so it does not require to have the `openmc.lib` enabled.

I don't quite see how we can do this without using the `openmc.lib` module at some point myself, but you have a way to accomplish this in mind, great!

pshriwise replied:
> Do you mean when using the `diff_burnable_mats` ? because we already write the override in the `cell.fill`. For `DAGMCCell`, it is then used to generate the `material_override` node in the xml.

Ah, yeah you're right. I was thinking that if the information is stored on the `DAGMCCell` objects then we're kind of duplicating it in the `material_overrides` attribute. So I was wondering if we might just populate the XML element from the `DAGMCCell` objects during export and not have an attribute on the `DAGMCUniverse`. However, I'm not sure how one would handle importing an XML file without generating `DAGMCCell` objects. I think we can look into that further down the line. Not super important at the moment.
 
> As to have `find_buranle_mats` only works with `DAGMCCell` when `openmc.lib` is enable, we could pass back the information directly to the cpp DAGMCCell. Or add a method to pass it.
> 
> Is that what you meant ?

This would be kind of nifty, but I don't think we need to worry about it at the moment.

bam241 replied:
I added a new implementation for this.

basically in the python DAGMCUniverse before generating the xml node for the DAGMCUniverse, one update the override_materials dict with the cell.fill (if the cells have multiple instance and an `Iterable` fill, i.e. the DAGMCCell is inserted multiple times in the Geometry, and has a different fill for each instance.)

If an override_material has already been set overloading per material name, I decided to keep both.
And on loading the DAGMC geometry (C++) to prioritise the override per cell Id then  override per material name then fall back to material assignment as stated in the DAGMC file.

pshriwise commented on src/dagmc.cpp:
```diff
@@ -816,12 +850,50 @@ int32_t next_cell(int32_t surf, int32_t curr_cell, int32_t univ)
+  const auto& univ = model::universes[model::universe_map[univ_id]];
+  if (univ->geom_type() != GeometryType::DAG) {
+    fatal_error(
+      "Universe " + std::to_string(univ_id) + " is not a DAGMC Universe!");
```
No need to include the `!` in the error messages. If they are in other messages added here, let's remove them.

```suggestion
      "Universe " + std::to_string(univ_id) + " is not a DAGMC Universe");
```

pshriwise reviewed (changes requested):
Another round. Getting closer I think!

bam241 commented:
@pshriwise I think I have addressed all your comments.

pshriwise commented on src/cell.cpp:
```diff
@@ -252,12 +252,12 @@ void Cell::to_hdf5(hid_t cell_group) const
 CSGCell::CSGCell()
 {
-  geom_type_ = GeometryType::CSG;
+  geom_type() = GeometryType::CSG;
```
There are a couple usages of `geom_type_` directly in `particle.cpp` (and perhaps elsewhere). Can you update those lines as well please?

bam241 replied:
changed all `geom_type_` call to `geom_type()`
change `suface.h` accordingly

pshriwise commented on openmc/lib/dagmc.py:
```diff
@@ -0,0 +1,62 @@
+        n
+    )
+    return cell_ids
+
```
[PEP8](https://peps.python.org/pep-0008/#blank-lines): Two lines between file-level functions 
```suggestion


```

pshriwise commented on openmc/model/model.py:
```diff
@@ -324,6 +325,25 @@ def init_lib(self, threads=None, geometry_debug=False, restart_file=None,
+        """
+        if self.is_initialized:
+            if self.materials:
+                mats = self.materials
```
mats --> materials

pshriwise commented on openmc/model/model.py:
```diff
@@ -1142,38 +1162,63 @@ def differentiate_depletable_mats(self, diff_volume_method: str):
             volume equally between the new materials, 'match cell' sets the
             volume of the material to volume of the cell they fill.
```
Let's include a description of the behavior if `diff_volume_method` is `None` (no action)

pshriwise commented on openmc/model/model.py:
```diff
@@ -1142,38 +1162,63 @@ def differentiate_depletable_mats(self, diff_volume_method: str):
+    def differentiate_mats(self, diff_volume_method: str = None, depletable_only: bool = True):
+        """Assign distribmats for each depletable material
+
+        .. versionadded:: 0.14.0
```
We'll need to update the version here at this point now that we're at `0.15.1-dev` now.

pshriwise commented on openmc/model/model.py:
```diff
@@ -1142,38 +1162,63 @@ def differentiate_depletable_mats(self, diff_volume_method: str):
+
+        Parameters
+        ----------
+        diff_volume_method : str
```
Missing documentation for `depletable_only` here.

pshriwise commented on openmc/model/model.py:
```diff
@@ -1142,38 +1162,63 @@ def differentiate_depletable_mats(self, diff_volume_method: str):
+            volume equally between the new materials, 'match cell' sets the
+            volume of the material to volume of the cell they fill.
+        """
+        if diff_volume_method not in ("divide equally", "match cell", None):
```
Let's use the `check_value` function here to verify the volume method. It will need to be included at the top of this file:

```python
from .checkvalue import check_value
```

pshriwise commented on openmc/model/model.py:
```diff
@@ -1142,38 +1162,63 @@ def differentiate_depletable_mats(self, diff_volume_method: str):
         # Count the number of instances for each cell and material
         self.geometry.determine_paths(instances_only=True)
 
         # Extract all depletable materials which have multiple instances
```
This comment should be updated based on the new behavior of the lines below.

pshriwise commented on openmc/model/model.py:
```diff
@@ -1142,38 +1162,63 @@ def differentiate_depletable_mats(self, diff_volume_method: str):
                                 )
                             cell.fill.volume = cell.volume
+                    if isinstance(cell, openmc.DAGMCCell):
+                        for i in range(cell.num_instances):
+                            cell.fill[i].name = f"{cell.fill[i].name}_{cell.id}_{i}"
```
```suggestion
                        for i, f in enumerate(cell.fill):
                            f.name += f"_{cell.id}_{i}"
```

pshriwise replied:
Follow-up to this. With the aim of keeping DAGMC-specific code out of otherwise generic blocks of code like this one, do you think it would make sense to populate these fill names during XML element creation for the DAGMC universe?

pshriwise replied:
Maybe in the `build_override_mat_from_cells` method?

pshriwise commented on openmc/cell.py:
```diff
@@ -780,3 +780,5 @@ def from_xml_element(cls, elem, surfaces, materials, get_universe):
         get_universe(univ_id).add_cell(c)
         return c
+
+
```
```suggestion
```

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,593 @@
+        # If the Cell is in the Universe's list of Cells, delete it
+        self._cells.pop(cell.id, None)
+
+    def sync_dagmc_cells(self, mats={}):
```
Should the `mats` parameter really be optional? It seems this method would fail without a valid entry.

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,593 @@
+    def sync_dagmc_cells(self, mats={}):
+        """Synchronize DAGMC cell information between Python and C API
+
+        .. versionadded:: 0.13.0
```
Need to update `versionadded` here as well.

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,593 @@
+                f"Material name '{mat_name}' not found in DAGMC file")
+        cv.check_iterable_type('material objects', overrides, str)
+
+        self.material_overrides[mat_name] = overrides
```
Does this support a single override as well as a list? Might be convenient to support that if one wants to override a material for all of the cell instances.

bam241 replied:
This should allow to override per cell_id and material name

pshriwise replied:
I don't quite understand why it's desirable to have both options. How do we resolve this if both the name of a material and a cell ID that is filled with that material are present in this dictionary?

pshriwise replied:
Did this get addressed? Can you explain how the above scenario will be handled?

bam241 replied:
I think I missed understood at first.

I think this is now addressed.

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,593 @@
+            None
+        """
+        for cell in self.cells.values():
+            if cell.n_instances > 1 and isinstance(cell.fill, Iterable):
```
Is there a reason this is restriced to cells with more than one instance?

bam241 replied:
your comment made me thinking :)

You are right, if `cell.fill` is an `Iterable` then `cell.n_instances` has to have more than one instance.


I though about the eventuallity to remove restrictions on `Iterable`,(and `cell.n_instance`), but as all DAGMCcells, when sync get their `fill` property that would me filling the override materials dict with all the fill of all DAGMCcells overridden or not.

pshriwise reviewed (changes requested):
This is looking good @bam241! We're getting into the nit-picky phase now. 

There are one or two sections that are DAGMC-specific in `model.py` and I wonder if we can figure out a way to build those into the DAGMC classes. I'll think on it some more, but then it'll fit in pretty seamlessly with native geometry in OpenMC.

bam241 commented:
@pshriwise I think I addressed your last round of comments.


I moved the DAGMCUniverse from `openmc.universe` to `openmc.dagmc` and shuffled some method from `Universe` to `UniverseBase` , and as `openmc.universe` has been updated on `develop`, it is difficult to make sure I didn't miss any update to `DAGMCUniverse`, `Universe` and `UniverseBase` while resolving the conflict.

I would suggest extra caution when reviewing those part, in case I missed any (I have been careful... but I might have missed something) and would ask for as much swiftness as we can have to conclude this (without impacting the quality) to avoid any additional complicated  conflict :)

bam241 commented:
@pshriwise here is an other illustration of the CI inconsistance:
https://github.com/bam241/openmc/pull/1

I have a PR with this branch on my fork, the `[Python 3.10 (omp=y, mpi=y, dagmc=y, ncrystal=, libmesh=, event= vectfit=)](https://github.com/openmc-dev/openmc/pull/3056#logs)`
test failed after 6h. while here it worked in 1h

pshriwise commented on include/openmc/dagmc.h:
```diff
@@ -184,6 +184,7 @@ class DAGUniverse : public Universe {
                              //!< generate new material IDs for the universe
   bool has_graveyard_; //!< Indicates if the DAGMC geometry has a "graveyard"
                        //!< volume
+  std::map<std::string, vector<std::string>> instance_material_overrides;
```
Let's add a doc string to this attribute please.

pshriwise commented on openmc/model/model.py:
```diff
@@ -324,6 +325,28 @@ def init_lib(self, threads=None, geometry_debug=False, restart_file=None,
 
+    def sync_dagmc_universe(self):
+        """
+        Synchronize all DAGMC universes with the current geometry.
```
```suggestion
        Synchronize all DAGMC universes in the current geometry.
```

pshriwise commented on openmc/model/model.py:
```diff
@@ -324,6 +325,28 @@ def init_lib(self, threads=None, geometry_debug=False, restart_file=None,
+
+        .. versionadded:: 0.15.1-dev
+
+        Returns:
+            None
+        """
```
No need to include this if there's no return value.

```suggestion
        """
```

pshriwise commented on openmc/universe.py:
```diff
@@ -135,6 +139,130 @@ def create_xml_subelement(self, xml_element, memo=None):
+    
+    def clear_cells(self):
+        """Remove all cells from the universe."""
+
```
```suggestion
```

pshriwise commented on src/dagmc.cpp:
```diff
@@ -211,15 +230,44 @@ void DAGUniverse::init_geometry()
       c->material_.push_back(MATERIAL_VOID);
     } else {
       if (uses_uwuw()) {
         uwuw_assign_material(vol_handle, c);
```
How do we want the overrides to work when UWUW materials are present? I think that we'd want them to override UWUW material assignments as well.

bam241 replied:
I did something about this:

first check if override is path by cell ID, 
then check if uwuw is used to recover the material name (as override key), 
finally use the mat_str as override key


in all 3 cases, use to legacy assignment to override mat assignment.

bam241 replied:
@pshriwise I left this one open so you can make sure this fit what you had in mind.

pshriwise commented on src/dagmc.cpp:
```diff
@@ -818,12 +866,50 @@ int32_t next_cell(int32_t surf, int32_t curr_cell, int32_t univ)
 
 namespace openmc {
 
+extern "C" int openmc_dagmc_universe_get_cell_ids(
```
These implementations should probably raise some kind of exception or throw a fatal_error when OpenMC is not configured to build with DAGMC support.

pshriwise commented on tests/unit_tests/dagmc/test_model.py:
```diff
@@ -0,0 +1,117 @@
+    mats["Water"].add_s_alpha_beta("c_H_in_H2O")
+    mats["Water"].name = "Water"
+
+    p = pkg_resources.resource_filename(__name__, "dagmc_differentiate_mat.h5m")
```
In the interest of minimizing the number of binary files in the repo and it's overall size, would one of the other existing DAGMC files in the regression tests work for this test instead?

pshriwise commented on src/dagmc.cpp:
```diff
@@ -818,12 +866,50 @@ int32_t next_cell(int32_t surf, int32_t curr_cell, int32_t univ)
+  *n = dag_cell_ids.size();
+}
+
+extern "C" int openmc_dagmc_universe_get_num_cells(int32_t univ_id, size_t* n)
```
We'll want all of these functions to return proper error codes to satisfy the C interface.

pshriwise commented:
Another set of comments for you @bam241. Sorry if they're a little hard to track -- should've done a proper review. 

This is looking quite good! Just a few spots to polish up and I think it'll be ready

bam241 commented:
@pshriwise I added an other test, to check more extensively what is done, I can probably factorise some code between the 2 tests, will do.

Not that I had to remove the test that checked that the number of instance of the cells matched the number of instance of the override when applying the material assignment override in `dagmc.cpp`.
The number of instances of the `Cell` is determine after loading the `DAGMC` model, therefore it is 0 when loading the `DAGMC` geometry... so we can't test against it at this moment...

bam241 commented:
I also, make an addition split method, to allow splitting based on an arbitrary `delim`, some material name (i.e. in the actual `dagmc.h5` used in the test) have space in it, therefore space is not a proper delimiter to separate the list of material names use for the override.
I used `;` instead

bam241 commented:
@pshriwise this is ready for a other round of review :)

pshriwise commented on openmc/model/model.py:
```diff
@@ -1154,51 +1175,85 @@ def update_material_volumes(self, names_or_ids, volume):
-        if diff_volume_method == 'divide equally':
+        if diff_volume_method == "divide equally":
             for mat in distribmats:
                 if mat.volume is None:
```
Let's move this check into it's own loop over `distribmats`. If error it hit for one material but not all, this will result in some Materials with volumes changed after an early exit, leaving their volume property in an invalid state.

pshriwise commented on openmc/model/model.py:
```diff
@@ -1154,51 +1175,85 @@ def update_material_volumes(self, names_or_ids, volume):
-                        for _ in range(cell.num_instances):
-                            cell.fill = mat.clone()
+                        cell.fill = mat.clone()
+                        for i in range(cell.num_instances):
```
Same suggestion here as the one above for materials (Sorry, I know this might be pre-existing). In fact, I can make an issue for this and address separately if you like.

pshriwise commented on openmc/model/model.py:
```diff
@@ -1154,51 +1175,85 @@ def update_material_volumes(self, names_or_ids, volume):
         distribmats = set(
             [mat for mat in self.materials
-                if mat.depletable and mat.num_instances > 1])
+                if (mat.depletable or not depletable_only) and mat.num_instances > 1])
```
Small thing, but let's give this interior expression it's own name for readability and a shorter set comprehension. Soon enough no one but you or I will be able to parse it. And even sooner we'll forget.

pshriwise commented on openmc/model/model.py:
```diff
@@ -1154,51 +1175,85 @@ def update_material_volumes(self, names_or_ids, volume):
-                    raise RuntimeError("Volume not specified for depletable "
-                                        f"material with ID={mat.id}.")
+                    raise RuntimeError(
+                        "Volume not specified for depletable "
```
Since we're supporting non-depletable materials in some cases now

```suggestion
                        "Volume not specified for "
```

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,617 @@
+        self.material_overrides = {}
+        for cell in self.cells.values():
+            if isinstance(cell.fill, Iterable):
+                self.material_overrides[str(cell.id)] = [mat.name for mat in cell.fill if mat]
```
Here the keys of the `material_overrides` are set to be the strings of cell IDs, but above material names are used. Here's my preferred key-value setup:

keys: Cell IDs (or `openmc.Cell` instances if we want to be fancy, and I think we do)
values: an `Iterable` of `openmc.Material` instances

pshriwise commented on src/dagmc.cpp:
```diff
@@ -818,12 +864,56 @@ int32_t next_cell(int32_t surf, int32_t curr_cell, int32_t univ)
+  std::vector<int32_t> dag_cell_ids;
+  for (const auto& cell_index : univ->cells_) {
+    const auto& cell = model::cells[cell_index];
+    if (cell->geom_type() == GeometryType::DAG)
```
The cells in this Universe should always be `DAGCell`'s. Maybe this could be a check to ensure that a `CSGCell` isn't found? Otherwise I think we can remove it.

pshriwise commented on src/dagmc.cpp:
```diff
@@ -72,6 +72,25 @@ DAGUniverse::DAGUniverse(pugi::xml_node node)
   }
 
+  // get material assignment overloading
+  if (check_for_node(node, "material_overrides")) {
```
We'll want to document the updates to the XML specification for `geometry.xml` to include this block in the `DAGMCUniverse` nodes.

pshriwise commented on include/openmc/dagmc.h:
```diff
@@ -184,6 +190,12 @@ class DAGUniverse : public Universe {
   bool has_graveyard_; //!< Indicates if the DAGMC geometry has a "graveyard"
                        //!< volume
+  std::map<std::string, vector<std::string>>
+    instance_material_overrides; ///!< Map of material overrides
```
More generally I think we can call this `material_overrides`. If an entry in this object is length 1, it can effectively an override for a cell without any additional instances.

pshriwise commented on src/dagmc.cpp:
```diff
@@ -609,14 +631,38 @@ void DAGUniverse::uwuw_assign_material(
+  // if Cell ID matches an override key, use it to override the material
+  // assignment else if UWUW is used, get the material assignment from the DAGMC
+  // metadata
+  std::cout << "XML instance"
+            << instance_material_overrides.count("id_" + std::to_string(c->id_))
+            << std::endl;
```
We probably don't want all of this output every time. Let's use the existing `message` functions and select an appropriate verbosity level for displaying this information.

pshriwise reviewed (changes requested):
With a few more updates I think this will be pretty much seamless! A few open questions to address yet as well.

Thanks for your effort on this. Certainly a non-trivial capability @bam241.

bam241 commented:
@pshriwise I think I addressed to all your questions or comments.

I left 1 unresolved comment, so you can "ok" how I have addressed your question about the `uwuw` assignment.

pshriwise commented on src/dagmc.cpp:
```diff
@@ -609,14 +630,39 @@ void DAGUniverse::uwuw_assign_material(
+  // if Cell ID matches an override key, use it to override the material
+  // assignment else if UWUW is used, get the material assignment from the DAGMC
+  // metadata
+  std::stringstream msg;
+  msg << "XML instance"
+      << material_overrides.count("id_" + std::to_string(c->id_));
+  write_message(msg.str(), 5);
```
I made use of `fmt` here as we do in other output messages and changed the verbosity level to 8. I don't think we want to see all of this information by default.

Can you incorporate this and update the content of the output message a bit. What information is meant to be provided to the user here?

```suggestion
  write_message(fmt::format("XML instance {}", c->id_), 8);
```

pshriwise commented on tests/unit_tests/dagmc/test_model.py:
```diff
@@ -0,0 +1,141 @@
+        ]
+    )
+
+    point = openmc.stats.Point((0, 0, 0))
+    source = openmc.IndependentSource(space=point)
```
This matches the default source's spatial description, so we don't need to specify it here.

```suggestion
    point = openmc.stats.Point((0, 0, 0))
    source = openmc.IndependentSource(space=point)
```

pshriwise commented on tests/unit_tests/dagmc/test_model.py:
```diff
@@ -0,0 +1,141 @@
+    source = openmc.IndependentSource(space=point)
+
+    settings = openmc.Settings()
+    settings.source = source
```
```suggestion
```

bam241 commented:
@pshriwise I have addressed your last comments, 

let me know if you have more :)

pshriwise commented on tests/unit_tests/dagmc/test_model.py:
```diff
@@ -0,0 +1,141 @@
+
+    daguniv = openmc.DAGMCUniverse(p,auto_geom_ids=True,)
+
+    def pattern(center, bc):
```
Did you have an objection to using a `RectLattice` here?

bam241 replied:
I have done it, just forgot to resolve it.

pshriwise commented on src/dagmc.cpp:
```diff
@@ -609,14 +630,37 @@ void DAGUniverse::uwuw_assign_material(
+  // metadata
+  // Notify User that an override is being applied on a DAGMCCell
+  write_message(fmt::format("Applying override for DAGMCCell {}", c->id_), 8);
+  if (material_overrides.count("id_" + std::to_string(c->id_))) {
```
To be honest, I'm not a fan of matching based on prefixes in the material names. We should be able to provide a direct mapping of which material should be applied in place of the one provided in the DAGMC file.

bam241 replied:
Sadly I can't set an attribute to an int.
what I did is add a "Cell_" prefix that I detect and remove when parsing the xml in the cpp"

pshriwise commented:
There's a lot of nice work here.  My remaining concern is the structure of the `material_overrides` attribute. I think this data structure should only accept Cell IDs as keys. The values can be a single `Material` instance or in the case that cell instances should be filled with different materials (or material clones), the value is an iterable of `Material` instances with length equal to the number of Cell instances. This provides a direct and natural mapping for how cell materials are assigned and is analogous to how this works for CSG cells in the XML files.

To me, using materials as keys forces this awkward check regarding how cells look up what material alias should be used (currently relying on special naming of materials). If there's an iterable of `Material`'s as a value the length of that entry limits is application to a specific set of cells and I can see us getting into trouble there.

What do you think @bam241? Am I missing something?

Either way, documentation should be updated to be consistent once we come to a conclusion on this.

bam241 commented:
> There's a lot of nice work here. My remaining concern is the structure of the `material_overrides` attribute. I think this data structure should only accept Cell IDs as keys. The values can be a single `Material` instance or in the case that cell instances should be filled with different materials (or material clones), the value is an iterable of `Material` instances with length equal to the number of Cell instances. This provides a direct and natural mapping for how cell materials are assigned and is analogous to how this works for CSG cells in the XML files.
> 
> To me, using materials as keys forces this awkward check regarding how cells look up what material alias should be used (currently relying on special naming of materials). If there's an iterable of `Material`'s as a value the length of that entry limits is application to a specific set of cells and I can see us getting into trouble there.
> 
> What do you think @bam241? Am I missing something?
> 
> Either way, documentation should be updated to be consistent once we come to a conclusion on this.

The only issue I see with that is the obligation to use the openmc.lib.

if we allow using the dogma material name assignment as overriding key, it would allow user to override materials assignment manually without getting the DAGMC Cell ID.

If we make the openmc.lib mandatory for this feature then we could allow user to pass it by name to the interface, and search to which DAGMC Cell it corresponds, so the user don't have to do the search by himself

bam241 commented:
I let you decide if making openmc.lib mandatory is ok or not.

pshriwise commented:
> I let you decide if making openmc.lib mandatory is ok or not.

I'm good with features that rely on `openmc.lib`. Our plotter capabilities rely heavily on this currently and so do the depletion capabilities of the `Model` class, so there's a precedent for that. In the long term this probably isn't ideal and we should move to a Python API that connects more seamlessly to the C++ library, but I don't think we should draw the line at this capability by any means.

bam241 commented:
> > I let you decide if making openmc.lib mandatory is ok or not.
> 
> I think it's reliance on `openmc.lib` is okay. Our plotter capabilities rely heavily on this currently and so do the depletion capabilities of the `Model` class, so there's a precedent for that. In the long term this probably isn't ideal and we should move to a Python API that connects more seamlessly to the C++ library, but I don't think we should draw the line at this capability by any means.


@pshriwise I think I have addressed all your comments.


I refactored the python side a bit, to force all the addition to the override dictionary to go through the add function, where all the checks are.

You now can make an addition to the override dictionary via:
- material_name => check if name is in the DAGMC Geometry => get all corresponding Cell and add them as key
- cell_id => Check if the CellID is in the DAGMC Geometry => Add the corresponding Cell as a single Key
- Cell => Check if the Cell is in the DAGMC Geometry => Add the Cell as a single key


To deal with the capability to override a Material by a single Material, if a single material is provided to the `add_override` method,  it is embedded in a `List` to keep the Iterrable format which simplifies the checking the type of the item in the list and building the xml file.

bam241 commented:
Also I had to keep a prefix for the xml attribute as it seems that an attribute cannot be a `int`.
so I kept a "Cell_" prefix which I remove on parsing the xml in the cpp.

pshriwise commented:
> Also I had to keep a prefix for the xml attribute as it seems that an attribute cannot be a `int`.
> so I kept a "Cell_" prefix which I remove on parsing the xml in the cpp.

Can you elaborate on this for my sake? They cannot be `int`'s, true, but why not convert the integer to a string without the prefix?

pshriwise commented:
> > > I let you decide if making openmc.lib mandatory is ok or not.
> > 
> > 
> > I think it's reliance on `openmc.lib` is okay. Our plotter capabilities rely heavily on this currently and so do the depletion capabilities of the `Model` class, so there's a precedent for that. In the long term this probably isn't ideal and we should move to a Python API that connects more seamlessly to the C++ library, but I don't think we should draw the line at this capability by any means.
> 
> @pshriwise I think I have addressed all your comments.
> 
> I refactored the python side a bit, to force all the addition to the override dictionary to go through the add function, where all the checks are.
> 
> You now can make an addition to the override dictionary via:
> 
> * material_name => check if name is in the DAGMC Geometry => get all corresponding Cell and add them as key
> * cell_id => Check if the CellID is in the DAGMC Geometry => Add the corresponding Cell as a single Key
> * Cell => Check if the Cell is in the DAGMC Geometry => Add the Cell as a single key
> 
> To deal with the capability to override a Material by a single Material, if a single material is provided to the `add_override` method, it is embedded in a `List` to keep the Iterrable format which simplifies the checking the type of the item in the list and building the xml file.

Nice! Yeah I like this. There's something to be said for wanting to replace material A in the DAGMC file with material B. Just to clarify, the entries in the data structure are still `Cell`'s/Cell IDs as keys and a list of `Material`'s / Material IDs as values?

bam241 commented:
In the data structure yes: if material A is assigned initially in Cell ids [X,Y,Z]
the data structure will look like:
`{"Cell_X":[B], "Cell_Y":[B], "Cell_Z":[C]}`, The "Cell_" prefix is automatically taken out when parsing the XML file on the cpp side. As the Cell ids are attribute in the XML file it seems the they can't be int only( I tried to add them as `str(id)` but it apparently wanted some chars...)

The user could use the following method to apply the replacement:

`DAGMCUniverse->add_material_override(key="A",override="B")`

pshriwise commented:
> In the data structure yes: if material A is assigned initially in Cell ids [X,Y,Z] the data structure will look like: `{"Cell_X":[B], "Cell_Y":[B], "Cell_Z":[B]}`,
> 
> _The "Cell_" prefix is automatically taken out when parsing the XML file on the cpp side. As the Cell ids are attribute in the XML file it seems the they can't be int only( I tried to add them as `str(id)` but it apparently wanted some chars...)_

I'm a little confused by this, because we currently write cell IDs as XML node attributes

https://github.com/openmc-dev/openmc/blob/fbb115921c71997d37e62748a077bdd369667786/openmc/cell.py#L648-L649

Maybe we aren't thinking about the same thing here. To be specific what I'm thinking is that the `material_overrides` node would look something like:

```xml
<dagmc_universe filename="dagmc.h5m" id="1">
  <material_overrides>
    <cell id="1">
      10 11 12
    </cell>
    ...
  </material_overrides>
</dagmc_universe>
```

And so in the Python data structure it would look like:

```Python
>>> print(dagmc_universe.material_overrides)

{1: [10, 11, 12], ...}
```

> The user could use the following method to apply the replacement: `mat_B = openmc.Material()` `DAGMCUniverse->add_material_override(key="A",override=mat_B)` or `DAGMCUniverse->add_material_override(key=Cell_X,override=mat_B)` `DAGMCUniverse->add_material_override(key=Cell_Y,override=mat_B)` `DAGMCUniverse->add_material_override(key=Cell_Z,override=mat_B)` or `DAGMCUniverse->add_material_override(key=Cell_X.id,override=mat_B)` `DAGMCUniverse->add_material_override(key=Cell_Y.id,override=mat_B)` `DAGMCUniverse->add_material_override(key=Cell_Z.id,override=mat_B)`
> 
> Value of the override can be Material or material Ids in the override_materials, and Material names, or materials Ids in the xml file, (default in the XML file would be the Material Ids).

Nice. Yeah I like this option for a material --> material override 👍🏻

bam241 commented:
@pshriwise I reshaped the xml layout.
it now looks like:
```
    <material_overrides>
      <cell id="8" material="12;13;14;15"/>
      <cell id="9" material="16;17;18;19"/>
      <cell id="10" material="20;21;22;23"/>
    </material_overrides>
```

I kept the ";" separator between material values, as material name could be passed, so I thought semicolon was better than space.
Le me know.

pshriwise commented:
> I kept the ";" separator between material values, as material name could be passed, so I thought semicolon was better than space.

I didn't catch that we might write material names to XML in our conversation above, sorry. Any reason not to limit the values of the overrides data structure to Material objects or IDs? If we have the name of a material, presumably the Material object is right there and we can pass it instead?

From

```python
dagmc_univ.add_material_override(cell_a.id, mat.name)
```

To

```python
dagmc_univ.add_material_override(cell_a.id, mat)
```

It's nice to be able to reference materials by name when marking up the DAGMC file in a CAD program, but the OpenMC Material object (partially by nature of having a name attribute) is equally rich.

bam241 commented:
> didn't catch that we might write material names to XML in our conversation above, sorry. Any reason not to limit the values of the overrides data structure to Material objects or IDs? If we have the name of a material, presumably the Material object is right there and we can pass it instead?

Not particularly, as eventually the override materials is assigned using the `legacy_assign_material` which allows both assignment by name or by id. I thought it could be nice to support both.

I switched from `;` separator the format to space separated. and updated the documentation accordingly.

so now xml looks like
```
    <material_overrides>
      <cell id="8" material="12 13 14 15"/>
      <cell id="9" material="16 17 18 19"/>
      <cell id="10" material="20 21 22 23"/>
    </material_overrides>
```

pshriwise commented on include/openmc/dagmc.h:
```diff
@@ -184,6 +188,11 @@ class DAGUniverse : public Universe {
                              //!< generate new material IDs for the universe
   bool has_graveyard_; //!< Indicates if the DAGMC geometry has a "graveyard"
                        //!< volume
+  std::map<int, vector<std::string>>
```
Now that we're only writing IDs to the XML file can we make the values here `vector<int>`?

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,624 @@
+        """
+        keys = []
+        if isinstance(key, str):
+            if key not in self.material_names:
+                raise ValueError(
+                    f"Material name '{key}' not found in DAGMC file")
+            elif int(key) in self.cells:
+                keys = [self.cells[int(key)]]
```
I don't think we should require that the key meet both of these conditions. It should be one or there other right?

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,624 @@
+            dagmc_element.append(mat_element)
+        xml_element.append(dagmc_element)
+
+    def build_overide_mat_from_cells(self):
```
```suggestion
    def build_override_mat_from_cells(self):
```

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,624 @@
+        Returns:
+            None
+        """
+        for cell in self.cells.values():
+            if isinstance(cell.fill, Iterable):
+                self.add_material_override(cell, [mat for mat in cell.fill if mat])
```
Shouldn't an entry be added even if the `Cell`'s fill isn't an `Iterable`?

pshriwise commented on openmc/geometry.py:
```diff
@@ -402,6 +402,23 @@ def get_all_nuclides(self) -> list[str]:
+        """
+        universes = self.get_all_universes()
+        dag_universes = {}
+        for id, uni in universes.items():
+            if isinstance(uni, openmc.DAGMCUniverse):
+                dag_universes[id] = uni
```
```suggestion
        for id, univ in universes.items():
            if isinstance(uni, openmc.DAGMCUniverse):
                dag_universes[id] = univ
```

pshriwise commented on openmc/model/model.py:
```diff
@@ -324,6 +325,26 @@ def init_lib(self, threads=None, geometry_debug=False, restart_file=None,
+            if self.materials:
+                materials = self.materials
+            else:
+                materials = self.geometry.get_all_materials().values()
```
```suggestion
                materials = list(self.geometry.get_all_materials().values())
```

pshriwise commented on openmc/model/model.py:
```diff
@@ -324,6 +325,26 @@ def init_lib(self, threads=None, geometry_debug=False, restart_file=None,
+                materials = self.materials
+            else:
+                materials = self.geometry.get_all_materials().values()
+            for dagmc_universe in self.geometry.get_all_dagmc_universes().values():
+                dagmc_universe.sync_dagmc_cells(materials)
```
This is the only place that `get_all_dagmc_universes` is called. I wonder if doing the DAGMCUniverse check here wouldn't suffice.

```suggestion
            for univ in self.geometry.get_all_universes().values():
                if isinstance(univ, openmc.DAGMCUniverse):
                    univ.sync_dagmc_cells(materials)
```

pshriwise reviewed (changes requested):
Another round of tidy-ing. Nearly there!

bam241 commented:
@pshriwise ready of an other round :)

pshriwise commented on tests/unit_tests/dagmc/test_model.py:
```diff
@@ -0,0 +1,116 @@
+    reason="DAGMC CAD geometry is not enabled.")
+
+
+def set_dagmc_model():
```
Let's turn this into a `pytest.fixture` that can be passed to the test functions below.

pshriwise commented on tests/unit_tests/dagmc/test_model.py:
```diff
@@ -0,0 +1,116 @@
+
+
+def test_model_differentiate_with_DAGMC(): 
+    model = set_dagmc_model()
```
Given our catch regarding the import of the `material_overrides` from XML we should probably add a test that this property can be correctly written to and read from the XML files.

pshriwise commented on tests/unit_tests/dagmc/test_model.py:
```diff
@@ -0,0 +1,116 @@
+def test_model_differentiate_depletable_with_DAGMC():
+    model = set_dagmc_model()
+
+    p = Path("differentiate_depletable_mats/divide_equally")
+    p.mkdir(parents=True, exist_ok=True)
```
The `run_in_tmpdir` fixture should take care of this for us.

pshriwise commented on tests/unit_tests/dagmc/test_model.py:
```diff
@@ -0,0 +1,116 @@
+    return model
+
+
+def test_model_differentiate_depletable_with_DAGMC():
```
```suggestion
def test_model_differentiate_depletable_with_DAGMC(run_in_tmpdir):
```

pshriwise commented on src/string_utils.cpp:
```diff
@@ -71,6 +71,31 @@ vector<std::string> split(const std::string& in)
   return out;
 }
 
+vector<std::string> split(const std::string& in, char delim)
```
Now that we're back to space-delimited arrays in the XML file, I think this function can be returned to its original state.

pshriwise commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,630 @@
+        """
+        for cell in self.cells.values():
+            if isinstance(cell.fill, Iterable):
+                self.add_material_override(cell, [mat for mat in cell.fill if mat])
```
Won't this alter the intended cell fills if some of the entries in the `cell.fill` are skipped?

bam241 commented on openmc/dagmc.py:
```diff
@@ -0,0 +1,620 @@
+        for cell in self.cells.values():
+            if cell.fill is None:
+                continue
+            fill = cell.fill if isinstance(cell.fill, openmc.Iterable) else [cell.fill]
```
```suggestion
            fill = cell.fill
```
This should be enough as `add_material_override()` ensures an Iterable to be added to the dictionary.

bam241 replied:
cc @pshriwise

pshriwise replied:
Agreed 👍🏻

pshriwise reviewed (approved):
I think this is in a good place @bam241! Thank you for all your effort on this!

paulromano reviewed (approved):
Just finished making some updates throughout. Thanks for your work on this @bam241!
